File: core/src/org/rexellentgames/dungeon/util/Animation.java
package org.rexellentgames.dungeon.util;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.utils.JsonReader;
import com.badlogic.gdx.utils.JsonValue;
import org.rexellentgames.dungeon.assets.Graphics;
import org.rexellentgames.dungeon.net.Network;

import java.util.ArrayList;
import java.util.HashMap;

public class Animation {
	private HashMap<String, ArrayList<Frame>> frames = new HashMap<>();

	public Animation(String file) {
		this(file, "");
	}

	public Animation(String file, String add) {
		JsonReader reader = new JsonReader();
		JsonValue root = reader.parse(Gdx.files.internal("sprites_split/" + file + ".json"));

		JsonValue meta = root.get("meta");
		JsonValue frameTags = meta.get("frameTags");
		JsonValue frames = root.get("frames");

		for (JsonValue tag : frameTags) {
			int from = tag.getInt("from");
			int to = tag.getInt("to");
			String state = tag.getString("name");

			ArrayList<Frame> framesList = new ArrayList<>();

			for (int i = from; i <= to; i++) {
				JsonValue frame = frames.get(i);
				String name = frame.name;
				int delay = frame.getInt("duration");

				// this: actor_towelknight 0.ase -- and state dead
				// into this: actor-towelknight-dead-00

				name = name.replace(".ase", "");
				name = name.replace('_', '-');
				name = name.replace(' ', '-');

				name = name.substring(0, name.length() - (Character.isDigit(name.charAt(name.length() - 2)) ? 3 : 2));
				name += add + "-" + state + "-" + String.format("%02d", i - from);

				framesList.add(new Frame(Graphics.getTexture(name), delay * 0.001f));
			}

			this.frames.put(state, framesList);
		}
	}

	public static Animation make(String file) {
		if (Network.SERVER) {
			return null;
		} else {
			return new Animation(file);
		}
	}

	public static Animation make(String file, String add) {
		if (Network.SERVER) {
			return null;
		} else {
			return new Animation(file, add);
		}
	}

	public AnimationData get(String state) {
		ArrayList<Frame> data = this.frames.get(state);

		if (data == null) {
			return null;
		}

		return new AnimationData(data);
	}

	public class Frame {
		public TextureRegion frame;
		public float delay;

		public Frame(TextureRegion frame, float delay) {
			this.frame = frame;
			this.delay = delay;
		}
	}
}

File: BurningKnight/Util/Animations/Animation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using BurningKnight.Assets.Graphics;
using BurningKnight.Util.Files;
using MoonSharp.Interpreter;
using Newtonsoft.Json;

namespace BurningKnight.Util.Animations
{
	[MoonSharpUserData]
	public class Animation
	{
		private Dictionary<string, List<Frame>> frames = new Dictionary<string, List<Frame>>();

		public Animation(string aname, string add = "")
		{
			if (add == null)
			{
				add = "";
			}
			
			FileHandle handle = FileHandle.FromRoot("Animations/" + aname + ".json");

			if (!handle.Exists())
			{
				Log.Error("Animation " + aname + " does not exist!");
				return;
			}

			AnimationHelper root = JsonConvert.DeserializeObject<AnimationHelper>(handle.ReadAll());

			foreach (var tag in root.Meta.FrameTags)
			{
				int from = (int) tag.From;
				int to = (int) tag.To;
				string state = tag.Name;
				List<Frame> frameList = new List<Frame>();

				for (int i = from; i <= to; i++)
				{
					var pair = root.Frames.ElementAt(i);
					FrameValue frame = pair.Value;
					
					string name = pair.Key;

					// this: actor_towelknight 0.ase -- and state dead
					// into this: actor-towelknight-dead-00

					name = name.Replace(".aseprite", "");
					name = name.Replace(".ase", "");
					name = name.Replace('_', '-');
					name = name.Replace(' ', '-');

					name = name.Substring(0, name.Length - (Char.IsDigit(name[name.Length - 2]) ? 3 : 2));
					name += add + "-" + state + "-" + (i - from).ToString("D2");
					
					frameList.Add(new Frame(Graphics.GetTexture(name), frame.Duration * 0.001f));
				}

				frames[state] = frameList;
			}
		}

		public AnimationData Get(string id)
		{
			if (frames.ContainsKey(id))
			{
				return new AnimationData(frames[id]);
			}

			return null;
		}
		
		public class Frame
		{
			public TextureRegion frame;
			public float delay;
			public float initial;

			public Frame(TextureRegion frame, float delay) {
				this.frame = frame;
				this.delay = delay;
				initial = delay;
			}
		}
	}
}